feat: ✨ bring in content on commits and PRs from elsewhere (#157)
seedcase-project/guidebook: workflows/commits.qmd

Reviewer: Could you also delete the text from where you pulled it in those other repos?

# Commits

This guide covers how we in the Seedcase Project commit changes to our
repositories. Following these guidelines makes changes easier and faster

> to review and creates a clear and consistent history. It also allows us

Reviewer: ```suggestion
to review and creates a clear and consistent history. It also allows for
```

Since it's a guide, the audience is someone learning or wanting to know a particular thing, so the pronouns should be "you", not "us" or "we"

> to automatically generate changelogs and increase the version number of

Reviewer: ```suggestion
automatically generate changelogs and increase the version number of
```

# Commits

This guide covers how we in the Seedcase Project commit changes to our

> repositories. Following these guidelines makes changes easier to review

Reviewer 2: ```suggestion
repositories. Following these guidelines makes changes easier and faster to review
```

and creates a clear and consistent history. It also allows us to
automatically generate changelogs and increase the version number of a
project.

This page describes when and what to commit, the Conventional Commits
specification, and how to apply it to both code and non-code projects.

## When and what to commit

Each commit should be a useful checkpoint that another person can
understand. In general, aim to:

- **Make atomic commits**: Each commit should represent one logical
  change. For example, update the same file path across several files in
  one commit without including an unrelated text edit.
- **Commit often**: Make commits frequently enough to capture meaningful
  progress and create useful points to return to.
- **Write meaningful messages**: Summarise what changed and, when it is

>   not obvious, explain why. Follow the

Reviewer 2: ```suggestion
  not obvious, explain why it was changed. Follow the
```

  [Conventional Commits](#conventional-commits) specification.

Atomic does not necessarily mean changing only one file or a few lines.
A commit can include several files when all their changes are necessary
for the same purpose. Conversely, unrelated changes in one file should

> usually be separate commits.

Reviewer 2: ```suggestion
be committed separately.
```

### Review and stage changes

Before committing, review the changes as if you were seeing them for the
first time. Check for accidental edits, generated files, debugging code,
secrets, and changes that belong in another commit. Also consider
whether the code or text could communicate its intent more clearly.

Stage only the changes that belong in the commit. In a terminal,
`git add --patch` (or `git add -p`) lets you review and stage one part
of a file at a time. You can do the same kind of selective staging and
review from the Source Control view in VS Code. This helps keep
individual commits scoped to one logical change, rather than adding
entire files which might contain several unrelated changes.

### Decide how much detail to include

Use the subject line for a concise summary. Add a body when a future
reader would need context that cannot be understood from the change
itself, such as the motivation, constraints, or reason for choosing one
approach over another. Focus the body on **why** the change was needed;

> the diff usually shows **what** changed.

Reviewer 2: If you agree on the comment below, maybe this could be moved to the conventional commits section that describes the commit message content and structure, maybe at around L103-106,? Then, this section could be removed entirely.

Author: Good idea!

Avoid putting the full history of a task in every commit. In our
pull-request workflow, information has different purposes:

- The issue describes the problem and discusses the proposed direction.
- The pull request describes the completed solution and gives reviewers
  the context they need.
- Commits describe the logical steps used to build that solution.
- The final squash commit records the completed change on `main`.

The [pull request guide](prs.qmd#writing-a-good-pr-description) explains

> what to include in a PR description.

Reviewer 2: This is a nice overview but I'm not sure it belongs here. Could be in an workflows index/overview page introducing our workflows and how they fit together?

Author: Hmm, yea I agree that would be more suitable but it seems like we don't use index pages in this guide; rather the section dividers are collapsible. Maybe we don't need this for the time being? I removed it here and created https://github.com/seedcase-project/guidebook/issues/159 to follow up.

- `<type>` is the type of change.
- `[optional scope]` indicates the area of the codebase affected by the
  change.
- `<description>` is a brief summary of the change (also called the
  "subject line").
- `[optional body]` provides additional context or details about the
  change.
- `[optional footer(s)]` includes breaking changes or references to
  issues/tasks.

The `<description>` section is a short summary of the change. Write it
in the imperative, present tense (i.e., "change" not "changed" nor
"changes") without a period at the end. Be specific enough that another
person can distinguish the commit from others without reading the diff.

::: callout-tip
In the Seedcase Project, we tend not to use `[optional scope]` too

> often.

Reviewer 2: ```suggestion
In the Seedcase Project, we don't use `[optional scope]`.
```
Right?

Reviewer 2: Maybe move this to the optional section below, where the other optional parts are described anyway?

Author: Moved and clarified!.

- `revert`: Reverts a previous commit.
- `chore`: Other changes that don't modify source or test files.

The `<description>` section is a short summary of the change. Write it

> in the imperative, present tense (i.e., "change" not "changed" nor

Reviewer: ```suggestion
in the imperative, present tense (i.e., "change" not "changed" or
```

"changes") without a period at the end. Be specific enough that another

> person can distinguish the commit from others without reading the diff.

Reviewer: ```suggestion
person can distinguish the commit from others without reading the difference.
```

::: callout-tip
It can be difficult to remember the available types and what they
represent. If you use VS Code, you can install the extension
[Conventional Commits](https://marketplace.visualstudio.com/items?itemName=vivaxy.vscode-conventional-commits)

> to help write messages adhering to the conventional commit format. This

Reviewer: ```suggestion
to help write messages adhering to the Conventional Commit format. This
```

They seem to use title case

extension also includes the emojis from Gitmoji. Adding emojis is a way
of making the messages more colorful.
:::

### Optional: scope, body, and footer(s)

> In the Seedcase Project, we don't use `[optional scope]`.

Reviewer: ```suggestion
::: callout-tip
In Seedcase, we haven't found a use for using `scope` in our work. For
larger, more complex repositories (e.g. monoliths) using `scope` is a good
way to communicate the change, but because we make smaller packages,
`scope` hasn't been very useful for us. When we initially tried using it, we
didn't end up having many `scope` tags and often used the same one or two.
So it just added redundant text to the commit message.
:::
```

Add an `[optional body]` when a future reader would need context that

> cannot be understood from the diff and `<description>` alone, such as

Reviewer: ```suggestion
cannot be understood from the Git difference and `<description>` alone, such as
```

the motivation, constraints, or reason for choosing one approach over

> another. Focus the body on **why** the change was needed; the diff

Reviewer: ```suggestion
another. Focus the body on **why** the change was needed; the change difference
```

usually shows **what** changed. It should be separated from the

> `<description>` with a blank line.

Reviewer: ```suggestion
`<description>` with a blank line.

::: callout-tip
Because we advocate and use a trunk-based development (every pull request is atomic and we regularly make pull requests) where each pull request is [squashed](prs.qmd#merging-a-pull-request) (more on that below), the `[optional body]` of commit messages isn't kept during 
the squash. For that reason, we also don't use `[optional body]` often.
:::
```

- `revert`: Reverts a previous commit.
- `chore`: Other changes that don't modify source or test files.

::: callout-tip
It can be difficult to remember the available types and what they

> represents. If you use VS Code, you can install the extension

Reviewer 2: ```suggestion
represent. If you use VS Code, you can install the extension
```

[Conventional Commits](https://marketplace.visualstudio.com/items?itemName=vivaxy.vscode-conventional-commits)
to help write messages ahdering to the conventional commit format. This
extension also includes the emojis from Gitmoji. Adding emojis is a nice

> way of making the messages more expressive and visually appealing.

Reviewer 2: Oh my, I think I wrote the "visually appealing" part at some point. Don't really like it, but maybe it's fine 😅

Author: Hehe, well let me know if you would like to change it. I usually don't use gitmojis so I don't really have anything to add about them myself and just kept whatever was in the doc already

Author: Edit, I did a small change focusing on that they add some color; feel free to suggest something else.

:::

### Optional: body and footer(s)

When the `<description>` is not enough, add an `[optional body]` that
provides the motivation and other context needed to understand the

> change. Separate it from the subject with a blank line. This might

Reviewer 2: ```suggestion
change. It should be separated from the `<description>` with a blank line. This might
```
To keep within the conventional commits language as much as possible.

> include the motivation for the change, how it was -implemented, or any

Reviewer 2: ```suggestion
include the motivation for the change, why it was implemented this way, or any
```
Bc isn't the "how" obvious from the diff?

> other relevant information that didn't fit in the -description.

Reviewer 2: ```suggestion
other relevant information that didn't fit in the `<description>`.
```

is committed, you also need to add a `!` after the `<type>` in the
commit message. For example:

```{.md}
feat!: new feature that breaks existing functionality

BREAKING CHANGE: This feature changes the way the existing functionality
works, so it is not backward compatible.
```

### Commits in a squash-merge workflow

We normally use [squash and merge](prs.qmd#merging-a-pull-request),

> which turns all commits from a PR into one commit on `main`. The

Reviewer: ```suggestion
which turns all commits from a pull request into one commit on `main`. The
```

Since it hasn't been introduced as PR here

individual branch commits should still be atomic and have meaningful
messages so that reviewers can follow the work. However, the final

> squash commit is the commit that must provide the durable,

Reviewer: ```suggestion
squash commit is the commit that must provide the Conventional Commit-compliant, durable,
```

is committed, you also need to add a `!` after the `<type>` in the
commit message. For example:

```{.md}
feat!: new feature that breaks existing functionality

BREAKING CHANGE: This feature changes the way the existing functionality
works, so it is not backward compatible.
```

### Commits in a squash-merge workflow

We normally use [squash and merge](prs.qmd#merging-a-pull-request),
which turns all commits from a PR into one commit on `main`. The
individual branch commits should still be atomic and have meaningful
messages so that reviewers can follow the work. However, the final
squash commit is the commit that must provide the durable,
release-relevant record.

Before merging, make sure the PR title follows the Conventional Commits
format because it becomes the subject of the squash commit. Review the
generated commit message as well: keep enough of the PR description to
explain the final change, remove template instructions and temporary

> review details, and retain relevant issue references.

Reviewer 2: Should this be in the PR doc since it's about our PR merge workflow and checking that the PR title follow the conventional commits convention?

Author: Good suggestion, it was mostly there already so we can remove this paragraph
